### PR TITLE
Fixing zoom issues AND fit data to bounds changed to aggregation

### PR DESCRIPTION
**Pull request by corrigancd**
This PR fixes 2 issues - 

**Issue 1 - https://github.com/sirensolutions/kibi-internal/issues/11538** 
The reason the !event.chart flag was in on previous commit (that caused other issues) was to handle a case where there was no data present on the map canvas (plus collar). The problem was that the uiState was not updated and the precision was not updated. In the example below

We see europe at the start
Then I zoom to the middle of the ocean where there is no data
Then I click fit data to bounds button

What used to happen was that the zoom and hence previous geohash precision was kept, this means that there would have been circle markers on the map that are too small. This PR fixes that by setting the uiState dynmaically to the map bounds that leaflet has set based on the fitting algorithm in databoundshelper.js 

HandlingCaseOfZoomingANDNoDataOnInitialMapExtent

**Issue 2 - https://github.com/sirensolutions/kibi-internal/issues/11537**
The first iteration of improving fit data to bounds was a document search which retrieved 500 documents. The observed issue with this was if you had just a few documents in one corner of the map and a much larger selection elsewhere meaning that it was possible for documents not to be retrieved in the small corner. This has been fixed by making a geohash aggregation request to ES meaning that ALL data within the selection will be accounted for. 

Note - this will need to be tested on large dataset like Skyhook

### CODE REVIEW

File: public/vislib/DataBoundsHelper.js [hunk 1]
```diff
@@ -1,22 +1,28 @@
...
+    const utils = require('plugins/enhanced_tilemap/utils');
 
     class BoundsHelper {
       constructor(params) {
         this.searchSource = params.searchSource;
         this.field = params.field;
       };
 
-      getBoundsOfEntireDataSelection() {
+      getBoundsOfEntireDataSelection(vis) {
```
Comment (szydan) [inline public/vislib/DataBoundsHelper.js]: Code LGTM 
Lets add a unit test for this method 
to 
1) check that the correct query (with aggregation) is sent
2) check if computed bounds are correct 

Should be doable with a bit of stubbing + using $httpBackend angular service to simulate es responses

File: public/vislib/DataBoundsHelper.js [hunk 2]
```diff
@@ -30,16 +29,15 @@ define(function () {
...
-        console.log(vis.aggs);
         const searchSource = new SearchSource();
-        searchSource.inherits(this.searchSource); // test callsWith <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
-        searchSource.filter(null); // test callsWith <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
-        searchSource.filter(queryFilter.getFilters()); // test callsWith <<<<<<<<<<<<<<<<<<<<<<<<<<<<<< []
-        searchSource.aggs(() => { // sinon.stub(SearchSource.prototype, 'aggs').callsFake(cb => someVariable = cb());
+        searchSource.inherits(this.searchSource);
+        searchSource.filter(null);
+        searchSource.filter(queryFilter.getFilters());
+        console.log(searchSource.aggs);
```
Comment (szydan) [inline public/vislib/DataBoundsHelper.js]: remove console.log

File: public/vislib/DataBoundsHelper.js [hunk 3]
```diff
@@ -30,16 +29,15 @@ define(function () {
...
+        searchSource.aggs(() => {
+          console.log('CALLED');
```
Comment (szydan) [inline public/vislib/DataBoundsHelper.js]: and here

File: public/vislib/__tests__/DataBoundsHelper.js [hunk 4]
```diff
@@ -2,16 +2,174 @@
...
+    schemas: {
+      metrics: [{ name: 'testmetricname' }],
+      all: {
+        byName: {
+          testmetricname: {}
+        }
+      }
+    },
+    responseConverter: (vis, table) => {
+
```
Comment (szydan) [inline public/vislib/__tests__/DataBoundsHelper.js]: empty line

File: public/vislib/__tests__/DataBoundsHelper.js [hunk 5]
```diff
@@ -2,16 +2,174 @@
...
+    }
+  },
+  isHierarchical: () => { },
+  aggs: [
+    {
+      id: "1",
```
Comment (szydan) [inline public/vislib/__tests__/DataBoundsHelper.js]: please use single quotes

File: public/vislib/__tests__/DataBoundsHelper.js [hunk 6]
```diff
@@ -20,68 +178,60 @@ describe('DataBoundsHelper', () => {
...
       SearchSource = Private(SearchSourceProvider);
       BoundsHelper = Private(require('plugins/enhanced_tilemap/vislib/DataBoundsHelper'));
-      console.log(SearchSource);
 
-      SearchSource.prototype.aggs = cb => aggs = cb();
-      // sinon.stub(SearchSource.prototype, 'aggs').callsFake(cb => aggs = cb());
-      sinon.stub(SearchSource.prototype, 'fetch').returns(Promise.resolve(fakeSearchResponse));
-    });
 
+      // SearchSource.prototype.aggs = cb => testAggs = cb();
```
Comment (szydan) [inline public/vislib/__tests__/DataBoundsHelper.js]: remove commented code

File: public/vislib/__tests__/DataBoundsHelper.js [hunk 7]
```diff
@@ -20,68 +178,60 @@ describe('DataBoundsHelper', () => {
...
-    //   }
-    // };
-
-
-    // console.log(respFromBoundsHelper);
 
+    const boundsHelper = new BoundsHelper(params);
+    sinon.stub(SearchSource.prototype, 'aggs').callsFake(cb => testAggs = cb());
+
+    // aggSpy = sinon.spy('aggs');
```
Comment (szydan) [inline public/vislib/__tests__/DataBoundsHelper.js]: and here

File: public/vislib/__tests__/DataBoundsHelper.js [hunk 8]
```diff
@@ -20,68 +178,60 @@ describe('DataBoundsHelper', () => {
...
+    // boundsHelper.searchSource.aggSpy()
+    //   .then(d => { console.log('edwin ', d); });
+
+    boundsHelper.getBoundsOfEntireDataSelection(fakeVis)
+      .then((entireBounds) => {
+
+        console.log(testAggs);
```
Comment (szydan) [inline public/vislib/__tests__/DataBoundsHelper.js]: and here

File: public/visController.js [hunk 9]
```diff
@@ -578,5 +553,72 @@ define(function (require) {
...
+
+    // saving checkbox status to dashboard uiState
+    map.map.on('overlayadd', function (e) {
+      $scope.vis.getUiState().set(e.name, e.name);
+    });
+    map.map.on('overlayremove', function (e) {
+      $scope.vis.getUiState().set(e.name, false);
+    });
+
+    map.map.on('moveend', _.debounce(function setZoomCenter(ev) {
```
Comment (szydan) [inline public/visController.js]: In separate PR can we rename this variable to something more meaningful please
like "map.lefletMap" or whatever the second map is as this is confusing that we have a map inside a map

File: .eslintrc.yml [hunk 10]
```diff
@@ -9,7 +9,7 @@ globals:
     angular: true
     jQuery: true
 rules:
-  import/no-unresolved: [ 'error', { ignore: [ 'plugins/enhanced_tilemap/', 'jquery', 'd3', '@elastic/', 'react', 'ui/', 'expect.js', 'sinon' ] }]
+  import/no-unresolved: [ 'error', { ignore: [ 'plugins/enhanced_tilemap/', 'jquery', 'd3', '@elastic/', 'react', 'ui/', 'expect.js', 'sinon', 'ng_mock' ] }]
```
[the remark this answers is not in the document]
Comment (corrigancd) [reply .eslintrc.yml]: I asked AJ about a `Unable to resolve path to module 'ng_mock'` lint error and he recommended adding it in.

File: public/vislib/DataBoundsHelper.js [hunk 11]
```diff
@@ -25,53 +31,43 @@ define(function () {
...
-            const topRight = L.latLng(maxLat, maxLon);
-            const bottomLeft = L.latLng(minLat, minLon);
-            return L.latLngBounds(topRight, bottomLeft);
+              });
+
+              const topRight = L.latLng(maxLat, maxLon);
+              const bottomLeft = L.latLng(minLat, minLon);
+              return L.latLngBounds(topRight, bottomLeft);
+            } else {
+              console.warn('Unable to fit bounds as no data was detected');
```
[the remark this answers is not in the document]
Comment (corrigancd) [reply public/vislib/DataBoundsHelper.js]: Is this for compatibility purposes? console.warn is used without the if in several parts of the code. If it should be changed throughout the app, I'll make a separate issue.

File: public/visController.js (hunk 9 above)
Comment (corrigancd) [reply public/visController.js]: Done - https://github.com/sirensolutions/kibi-internal/issues/11579

**Comment by corrigancd** [comment]
Included fix for https://github.com/sirensolutions/kibi-internal/issues/11582 just now also. It was 1 line which stops events taking place in rzSlider from bubbling onto the map and causing the race condition outlined. It's difficult to see from the GIF, but I double click and drag a bit and all seems good: 

StopPropagationOnLegendSlider

**Comment by szydan** [comment]
@edwin merge if ready please

**Pull request merged**